lusidtools/cocoon/cocoon.py

```diff
@@ -287,6 +294,80 @@ def load_instrument_property_batch(
 
         return results
 
+    @staticmethod
+    @run_in_executor
+    def load_portfolio_group_batch(
+        api_factory: lusid.utilities.ApiClientFactory,
+        portfolio_group_batch: list,
+        **kwargs,
+    ) -> lusid.models.PortfolioGroup:
+        """
+        Upserts a batch of portfolios to LUSID
+        :param lusid.utilities.ApiClientFactory api_factory: The api factory to use
+        :param list[lusid.models.CreateTransactionPortfolioRequest] portfolio_batch: The batch of portfolios to create
+        :param kwargs: 'scope', 'code' arguments required for the API call
+        :return: lusid.models.PortfolioGroup: The response from LUSID
+        """
+
+        updated_request = group_request_into_one(
+            portfolio_group_batch[0].__class__.__name__,
+            portfolio_group_batch,
+            ["values"],
+        )
+
+        if "scope" not in list(kwargs.keys()):
+            raise KeyError(
+                "You are trying to load a portfolio group without a scope, please ensure that a scope is provided."
+            )
+
+        if "code" not in list(kwargs.keys()):
+            raise KeyError(
+                "You are trying to load a portfolio group without a portfolio code, please ensure that a code is provided."
+            )
+
+        try:
+
+            current_portfolio_group = api_factory.build(
+                lusid.api.PortfolioGroupsApi
+            ).get_portfolio_group(scope=kwargs["scope"], code=kwargs["code"])
+
+            #  Capture all portfolios - the ones currently in group + the new ones to be added
+            all_portfolios_to_add = (
+                updated_request.values + current_portfolio_group.portfolios
+            )
+
+            # Parse out new portfolios only
+            new_portfolios = [
+                code
+                for code in all_portfolios_to_add
+                if code not in current_portfolio_group.portfolios
+            ]
+
+            for code, scope in set(
+                [(resource.code, resource.scope) for resource in new_portfolios]
+            ):
+                api_factory.build(lusid.api.PortfolioGroupsApi).add_portfolio_to_group(
+                    scope=kwargs["scope"],
+                    code=kwargs["code"],
+                    effective_at=datetime.now(tz=pytz.UTC),
```

Comment:
I have added error handling to the loop of updates.
The desired behaviour - the function should update the portfolio group with portfolios if possible. One bad **add_portfolio_to_group** request should not preclude the remainder of the update.

Reply:
With this change unfortunately it will have the opposite effect. When you return the error on the failed adding of a portfolio that is the end of the function and no more portfolios will be added. Your test test_08_pg_add_bad_portfolio asserts that with a single bad portfolio an error will be returned. Why don't you try a test with one bad and one good portfolio and see if the good portfolio goes in and the bad one does not?